protocols/request-response/tests/error_reporting.rs

```diff
@@ -0,0 +1,727 @@
+use async_std::channel;
+use async_std::future::timeout;
+use async_std::task::{sleep, spawn};
+use async_trait::async_trait;
+use futures::prelude::*;
+use libp2p_request_response as request_response;
+use libp2p_request_response::ProtocolSupport;
+use libp2p_swarm::{StreamProtocol, Swarm};
+use libp2p_swarm_test::SwarmExt;
+use request_response::{Codec, InboundFailure, OutboundFailure};
+use std::time::Duration;
+use std::{io, iter};
+
+#[async_std::test]
+async fn report_outbound_failure_on_read_response() {
+    let _ = env_logger::try_init();
+
+    let protocols = iter::once((StreamProtocol::new("/test/1"), ProtocolSupport::Full));
+    let cfg = request_response::Config::default();
+
+    let mut swarm1 = Swarm::new_ephemeral(|_| {
+        request_response::Behaviour::<TestCodec>::new(protocols.clone(), cfg.clone())
+    });
+    let peer1_id = *swarm1.local_peer_id();
+
+    let mut swarm2 =
+        Swarm::new_ephemeral(|_| request_response::Behaviour::<TestCodec>::new(protocols, cfg));
+    let peer2_id = *swarm2.local_peer_id();
+
+    swarm1.listen().await;
+    swarm2.connect(&mut swarm1).await;
+
+    // On panic `panic_check_rx` will be closed
+    let (panic_check_tx, panic_check_rx) = channel::bounded::<()>(1);
+
+    // Server
+    let swarm1_task = async move {
+        let _panic_check_tx = panic_check_tx;
+        let mut req_id = None;
+
+        loop {
+            match swarm1.select_next_some().await.try_into_behaviour_event() {
```

Comment:
The `SwarmExt` trait has a bunch of useful functions  that e.g. automatically apply timeouts etc. If you use those, I think you shouldn't need to use the `timeout` below.